### PR TITLE
Pathfinding code refactoring

### PR DESCRIPTION
**Pull request by ArseniyShestakov**
This one not fully ready, but I can already collect more feedback as well as get your thoughts on following ideas:
1. What about move CPathfinder and everything related into separate file?
2. What other code must belong to pathfinder?
There also several functions that may be better placed within pathfinder related code:
   * CGameState::getMovementCost is first candidate and it's probably need own class like CMovementCost. Reason for this is fact that it's may be used by external code with current hero state, but in pathfinding it's must calculate cost based on current pathfinding state. E.g boat.
   * getNeighbours / canMoveBetween / checkForVisitableDir / etc.

It's just ideas and I don't sure what exactly must be moved into pathfinder, but moving code out CGameState would make future development easier.

Once this code is merged I'm going to work on something bigger:
http://forum.vcmi.eu/viewtopic.php?t=1115
Then I can finally get rid of "useEmbarkCost" and move all cost calculations into separate function. For this branch it's not viable because flying / water walking require huge changes in same area anyway.

### CODE REVIEW

File: lib/CGameState.cpp [hunk 1]
```diff
@@ -3558,21 +3602,46 @@ bool CPathfinder::goodForLandSeaTransition()
...
+
 	return true;
 }
 
+CPathfinder::PathfinderOptions::PathfinderOptions()
+{
+	useFlying = false;
+	useWaterWalking = false;
+	useEmbarkAndDisembark = true;
+	useTeleportTWoWay = true;
```
Comment (alexvins) [inline lib/CGameState.cpp]: typo: "TWo"

File: lib/CGameState.h [hunk 2]
```diff
@@ -279,29 +279,52 @@ struct DLL_EXPORT DuelParameters
...
-	bool goodForLandSeaTransition(); //checks if current move will be between sea<->land. If so, checks it legality (returns false if movement is not possible) and sets useEmbarkCost
+	bool isMovementPossible(); //checks if current move will be between sea<->land. If so, checks it legality (returns false if movement is not possible) and sets useEmbarkCost
+
+	int3 getSourceGuardPosition();
+	bool isSourceGuarded();
+	bool isDestinationGuarded();
+	bool isDestinationGuardian();
+
+	void getNeighbours(const int3 &coord);
+	void getTeleportExits(bool noTeleportExcludes = false);
```
Comment (alexvins) [inline lib/CGameState.h]: "getXXX" is bad function name if it returns nothing.

File: lib/CGameState.h [hunk 3]
```diff
@@ -279,29 +279,52 @@ struct DLL_EXPORT DuelParameters
...
 
 	int3 curPos;
 	CGPathNode *cp; //current (source) path node -> we took it from the queue
 	CGPathNode *dp; //destination node -> it's a neighbour of cp that we consider
 	const TerrainTile *ct, *dt; //tile info for both nodes
 	ui8 useEmbarkCost; //0 - usual movement; 1 - embark; 2 - disembark
-	Obj destTopVisObjID;
 
+	CGObjectInstance *sTileObj;
+	CGObjectInstance *dTileObj;
```
Comment (alexvins) [inline lib/CGameState.h]: Why non const?

File: lib/CGameState.h [hunk 4]
```diff
@@ -279,29 +279,52 @@ struct DLL_EXPORT DuelParameters
...
+	const CGTeleport *sTileTeleport;
+	const CGTeleport *dTileTeleport;
```
Comment (alexvins) [inline lib/CGameState.h]: consider adding (inline) setter for [sd]TileObj to ensure that [sd]TileTeleport is relevant.

**Comment by ArseniyShestakov** [comment]
Related question: why don't we store "valueType" in something like "bonuses.json"?

This could be good idea for bonuses that certainly can only have specific "valueType" that shouldn't be overridden.

**Comment by DjWarmonger** [comment]
I don't see what you mean by that. You already cna configure valueType via bonus json structure. Then the way these are interpreted depends on program code.

**Comment by ArseniyShestakov** [comment]
What I meant: both bonuses I fixed used in both spells and artifacts. Each time bonus used you have to specify same "valueType" as otherwise result can be inconsistent.

So I think it's would be handy to be able to predefine "valueType" once.

**Comment by alexvins** [comment]
> why don't we store "valueType" in something like "bonuses.json"?
Because bonus system was designed for "complex" bonus types - for them ALL value types make sense.

**Comment by ArseniyShestakov** [comment]
Still in some cases that are not complex simple option can be helpful too.
After all override can be supported and likely it's what config reader do in similar cases.

Anyway I just thought it's may be already supported while I missed it.

**Comment by alexvins** [comment]
It is not such difficult to add default value type to bonus type config and use it on deserialization|construction but we need to find ALL places where current default value - ADDITIVE_VALUE - is used.

File: lib/CGameState.cpp [hunk 5]
```diff
@@ -3284,45 +3281,154 @@ void CPathfinder::initializeGraph()
 			for(size_t k=0; k < out.sizes.z; ++k)
 			{
 				curPos = int3(i,j,k);
```
Comment (alexvins) [inline lib/CGameState.cpp]: you can use curPos attributes itself as cycle counters.
```cpp
for(curPos.x=0; curPos.x < out.sizes.x; curPos.x++)
```

Comment (ArseniyShestakov) [inline lib/CGameState.cpp]: Nice idea. I wanted to get rid of "curPos" as class member anyway as I want to use evaluateAccessibility outside of initializeGraph (I don't need whole graph on server for path validation). Good reason to do that.

**Comment by ArseniyShestakov** [comment]
So I think I'm done with code improvements. I'm do dislike unsafe "sTileObj" (and agree with @alexvins on it), but I don't feel it's necessary to work on it now because code will be heavily rewritten anyway. Goal to this branch was mainly to make pathfinder logic we have now readable so it's easier to not miss anything important once I rewrite it.

Though I still wonder about moving everything into separate file. Any pros/cons?

File: lib/CGameState.cpp [hunk 6]
```diff
@@ -3276,53 +3273,162 @@ TeamState::TeamState()
...
+}
+
+int3 CPathfinder::getSourceGuardPosition()
+{
+	return gs->map->guardingCreaturePositions[cp->coord.x][cp->coord.y][cp->coord.z];
+}
+
+bool CPathfinder::isSourceGuarded()
+{
+	//TODO: find out why exactly source can't be guarded if hero on it. Why this quirk was nessesaary in first place? Map where hero start on guarded tile?
```
Comment (IvanSavenko) [inline lib/CGameState.cpp]: Hero starts on guarded tile, hero teleports on guarded tile (via teleporter), monsters spawned near hero on new month. This is normal situation and not some exception from the rules.

**Comment by IvanSavenko** [comment]
> Though I still wonder about moving everything into separate file. Any pros/cons?

+1 to readability (pathfinder.cpp is more logical place for pathfinder than gamestate.cpp)
-1 to speed when using ancient compilers without link-time optimization which are likely not supported by vcmi anyway due to c++11 requirements. Not really applicable for pathfinder since large functions are not likely to be inlined anyway.

So yeah, good idea. In general I'd advice splitting files where possible to avoid  those unmanagable 10k lines files. IMO all files with more than 1k lines of code are a solid candidates for a split.

File: lib/CGameState.cpp (hunk 6 above)
Comment (ArseniyShestakov) [inline lib/CGameState.cpp]: Yeah, that explained it. For some reason when I wrote that comment I forgot that H3 should always let you leave guarded tile, but not let you go through it.

**Comment by ArseniyShestakov** [comment]
I'm not sure it's absolutely needed, but it's clearly going to make my future work a lot easier because I can now only look on what I work on and function order is fixed. Considering pathfinder code have nothing to do with CGameState it's would be really weird decision to move blocks of code all around in that file.

BTW it's ready to merge, but I can't fix VS project files or test it on Windows it's likely going to be up to @DjWarmonger to merge when he have time to do so.

**Comment by ArseniyShestakov** [comment]
And yeah I'm decide to put it in this pull request because it's will be even more frustrating if diff of next pull request going to looks like that.

**Comment by alexvins** [comment]
My comment is about particular line. It just shows weird here - look to commit.

**Comment by ArseniyShestakov** [comment]
Ops. There is several other place in VCMI code where shorter struct declaration used (CGameStatem, CModHandler, etc) so I consider it's reasonable to use for struck that will only be used once.

**Comment by ArseniyShestakov** [comment]
It's won't compile without RegisterTypes and likely because some code use CGameState internals.
Though I all the same functionality available via callbacks so likely I can get rid of that.

**Comment by ArseniyShestakov** [comment]
I think it's better this way, but in future I'll try to avoid CGameState usage too as currently pathfinder duplicate few simple functions that already found in callback. Also already find out that in CGTeleport I pass whole gamestate while it's absolutely not needed.

**Pull request merged**